File: app/src/main/java/de/mchllngr/quickopen/receiver/NotificationStarterReceiver.java
package de.mchllngr.quickopen.receiver;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import de.mchllngr.quickopen.model.ApplicationModel;
import de.mchllngr.quickopen.service.NotificationService;
import timber.log.Timber;

/**
 * {@link BroadcastReceiver} for (re)starting the {@link NotificationService}.
 *
 * This abstract approach is needed, because each action needs it's own {@link BroadcastReceiver}.
 */
abstract class NotificationStarterReceiver extends BroadcastReceiver {

    /**
     * Returns true when the received action is correct for this {@link BroadcastReceiver}.
     *
     * @param action received action
     * @return true is the action is correct for this {@link BroadcastReceiver}
     */
    protected abstract boolean isCorrectAction(@Nullable String action);

    /**
     * Start the {@link NotificationService} if called with the correct actions.
     */
    @Override
    public void onReceive(Context context, Intent intent) {
        if (context == null || intent == null) return;

        String action = intent.getAction();
        if (isCorrectAction(action)) {
            Timber.d("Action '" + action + "' received");
            ApplicationModel.removeNotLaunchableAppsFromList(context);
            startNotificationService(context);
        } else
            Timber.w("Unknown action '" + action + "' received");
    }

    protected void startNotificationService(@NonNull Context context) {
        Intent serviceIntent = new Intent(context, NotificationService.class);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O)
            context.startForegroundService(serviceIntent);
        else
            context.startService(serviceIntent);
    }
}

File: app/src/main/java/de/mchllngr/quickero/receiver/NotificationStarterReceiver.kt
package de.mchllngr.quickero.receiver

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import dagger.hilt.android.AndroidEntryPoint
import de.mchllngr.quickero.repository.application.ApplicationsRepository
import de.mchllngr.quickero.service.NotificationService
import de.mchllngr.quickero.util.notification.NotificationHelper
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import timber.log.Timber
import javax.inject.Inject

/**
 * [BroadcastReceiver] for (re)starting the [NotificationService] called when
 * [Intent.ACTION_BOOT_COMPLETED] or [Intent.ACTION_MY_PACKAGE_REPLACED] is received.
 */
@AndroidEntryPoint
class NotificationStarterReceiver : BroadcastReceiver() {

    @Inject lateinit var applicationsRepository: ApplicationsRepository
    @Inject lateinit var notificationHelper: NotificationHelper

    override fun onReceive(
        context: Context?,
        intent: Intent?
    ) {
        if (context == null || intent == null) return

        val action = intent.action
        if (Intent.ACTION_BOOT_COMPLETED == action || Intent.ACTION_MY_PACKAGE_REPLACED == action) {
            Timber.d("Action '$action' received")
            GlobalScope.launch {
                applicationsRepository.removeInvalidApplications()
                notificationHelper.startNotificationService()
            }
        } else {
            Timber.w("Unknown action '$action' received")
        }
    }
}
